misc/springsystem/src/processors/springsystemprocessor.cpp

```diff
@@ -0,0 +1,286 @@
+        springIndices->add(static_cast<uint32_t>(spring.second));
+    }
+
+    nodeOutport_.setData(nodeMesh);
+    springOutport_.setData(springMesh);
+}
+
+void SpringSystemProcessor::updateSystemFromProperties() { springSystem_ = getSystem(); }
+
+GravitySpringSystem<2, double> SpringSystemProcessor::getSystem() {
+    meshDirty_ = true;
+
+    auto grid = [&]() {
+        switch (springLayout_.get()) {
+            default:
+                [[fallthrough]];
+            case 1: {
+                return springmass::createLineGrid<2, double>(nodeCount_, nodeSpacing_);
+            }
+            case 2: {
+                const double offset = (std::ptrdiff_t{nodeCount_} - 1) * nodeSpacing_ * 0.5;
+                const dvec2 origin(-offset, offset);
+                return springmass::createRectangularGrid<2, double>(size2_t(nodeCount_), origin,
+                                                                    dvec2(nodeSpacing_));
+            }
+            case 3: {
+                const double offset = (nodeCount_ - 1) * nodeSpacing_ * 0.5;
+                const dvec2 origin(-offset, offset);
+                return springmass::createDiagonalGridDiagonal<2, double>(
+                    size2_t(nodeCount_), origin, dvec2(nodeSpacing_));
+            }
+            case 4: {
+                const double offset = (nodeCount_ - 1) * nodeSpacing_ * 0.5;
+                const dvec2 origin(-offset, offset);
+                return springmass::createHexagonalGrid<2, double>(size2_t(nodeCount_), origin,
+                                                                  dvec2(nodeSpacing_));
+            }
+        }
+    }();
+
+    return GravitySpringSystem<2, double>{deltaT_,
+                                          std::move(grid.positions),
+                                          std::move(grid.springs),
+                                          std::move(grid.locked),
+                                          externalForce_,
+                                          nodeMass_,
+                                          springConst_,
+                                          springRestLength_,
+                                          dampingCoeff_};
+}
+
+void SpringSystemProcessor::handlePicking(PickingEvent* p) {
+    if (p->getPressState() == PickingPressState::Move &&
+        p->getPressItems().count(PickingPressItem::Primary)) {
+        const auto worldDelta = p->getWorldSpaceDeltaAtPressDepth(camera_);
+        const auto ind = p->getPickedId();
+
+        springSystem_.position(ind) += dvec2{worldDelta};
+        p->markAsUsed();
+        meshDirty_ = true;
+        invalidate(InvalidationLevel::InvalidOutput);
+    }
+    if (p->getPressState() == PickingPressState::Press &&
+        p->getPressItem() == PickingPressItem::Primary) {
+        springSystem_.lockedNodes[p->getPickedId()] = true;
+    }
+    if (p->getPressState() == PickingPressState::Release &&
```

Comment:
check for state == `updated` in particular here! Release events might be spawned for other Ids, which then inform this ID about hover enter.

Reply:
Dont follow this, can you explain?